src/main/kotlin/io/github/mojira/arisa/ArisaMain.kt

```diff
@@ -1,165 +1,68 @@
-                        config, moduleRegistries, getSearchIssues(jiraClient, config)
-                    )
-                    TimeUnit.SECONDS.sleep(checkIntervalSeconds)
-                } catch (exception: Exception) {
-                    log.error("Relog failed", exception)
-                    // Wait before performing any further action
-                    TimeUnit.SECONDS.sleep(max(RELOG_INTERVAL_IN_SECONDS, checkIntervalSeconds))
-                }
-            } else {
-                // Not enough time for relog passed, just try waiting a bit
-                log.info("Not enough time for relog has passed")
-                TimeUnit.SECONDS.sleep(max(WAIT_TIME_AFTER_CONNECTION_ERROR_IN_SECONDS, checkIntervalSeconds))
-            }
-        }
+        HelperMessageUpdateService.checkForUpdate()
 
-        // Update helper messages if necessary
-        if (currentTime.epochSecond - helperMessagesLastFetch.epochSecond >= helperMessagesInterval) {
-            HelperMessageService.updateHelperMessages(helperMessagesFile)
-            executor = Executor(
-                config, moduleRegistries, getSearchIssues(jiraClient, config)
-            )
-            helperMessagesLastFetch = currentTime
-        }
+        val secondsToSleep = runExecutionCycle(executor)
+        TimeUnit.SECONDS.sleep(secondsToSleep)
     }
 }
 
-private fun getSearchIssues(
-    jiraClient: JiraClient,
-    config: Config
-): (String, Int, () -> Unit) -> List<Issue> {
-    return ::searchIssues
-        .partially1(jiraClient)
-        .partially1(config)
-}
+/**
+ * @return amount of seconds to sleep after this execution cycle
+ */
+private fun runExecutionCycle(executor: Executor): Long {
+    val timeframe = ExecutionTimeframe.getTimeframeFromLastRun()
+    val currentRunTime = timeframe.currentRunTime
 
-private fun readLastRunTime(lastRun: List<String>): Instant {
-    return if (lastRun[0].isNotEmpty())
-        Instant.ofEpochMilli(lastRun[0].toLong())
-    else Instant.now().minus(TIME_MINUTES, ChronoUnit.MINUTES)
-}
+    // Execute all enabled modules using the executor
+    val executionResults = executor.execute(timeframe, LastRun.failedTickets)
+
+    return if (executionResults.successful) {
+        // Reset relog timer
+        connectionService.notifyOfSuccessfulConnection()
 
-private fun readLastRun(lastRunFile: File): List<String> {
-    return (if (lastRunFile.exists())
-        lastRunFile.readText().trim()
-    else "")
-        .split(",")
+        // Update last run and save it to file
+        LastRun.update(currentRunTime, executionResults.failedTickets)
+        if (config[Arisa.Debug.updateLastRun]) {
+            LastRun.writeToFile()
+        }
+
+        MIN_TIME_BETWEEN_EXECUTION_CYCLES_IN_SECONDS
+    } else {
+        connectionService.tryRelog().sleepTimeInSeconds
+    }
 }
 
+/**
+ * Reads the config from the config file(s) and other sources
+ */
 private fun readConfig(): Config {
```

Comment:
I would move all the config stuff to its own service too

Reply:
Hmm, do we need to? It's literally only this function.

Reply:
Ideally Main doesnt even need the config, so this can be its own service and we call it when we need it(multiple instances of it is fine)